Fix 32619: clarify font-weight descriptor (#32853)
mdn/content: files/en-us/web/css/@font-face/font-weight/index.md

## Examples

### Selecting normal and bold fonts

In this example we include two fonts, one normal weight, one bold weight, from the ["Fira Sans"](https://fonts.google.com/specimen/Fira+Sans) font family using two `@font-face` at rules. We set `font-weight` descriptors to match the weight of the fonts.

After this, CSS rules can select the normal or the bold font for the "Fira Sans" family just by setting the {{cssxref("font-weight")}} property. Note that the {{htmlelement("strong")}} HTML element also selects the bold font, because by default `<strong>` elements use a bold font.

#### HTML

```html
<p class="one">Fira Sans, normal weight paragraph</p>
<p class="two">Fira Sans, bold weight paragraph</p>

> <strong>Fira Sans, default weight &lt;strong&gt; element</strong>

Reviewer: ```suggestion
<p><strong>Fira Sans, default weight &lt;strong&gt; element</strong></p>
```

  font-size: 2rem;
}

p.one {
  font-weight: normal;
}

p.two {
  font-weight: bold;
}
```

#### Result

{{embedlivesample("Selecting normal and bold fonts", "", 300)}}

### Setting a range for a variable font

In this example we include a variable font, ["League Mono"](https://www.theleagueofmoveabletype.com/league-mono), using a single `@font-face` at-rule. We've used a `font-weight: 300 700` value to effectively limit the range of weights that are available. If a CSS rule uses our "League Mono" font, then if it specifies a weight outside this range the weight it gets is clamped to the range.

To show the effect of this we've included another font using "League Mono" that does not set the `font-weight` descriptor, and we've called this "League Mono-complete".

#### HTML

```html
<p class="one">League Mono-complete, font-weight 200</p>
<p class="two">League Mono, font-weight 400</p>
<p class="three">League Mono, font-weight 600</p>
<p class="four">League Mono-complete, font-weight 800</p>
```

#### CSS

```css
@font-face {
  font-family: "League Mono";
  src: url("https://mdn.github.io/shared-assets/fonts/LeagueMono-VF.ttf");
  font-weight: 300 700;
}

@font-face {

>   font-family: "League Mono-complete";
>   src: url("https://mdn.github.io/shared-assets/fonts/LeagueMono-VF.ttf");

Reviewer: ```suggestion
  font-family: "League Mono-complete";
  src: url("https://mdn.github.io/shared-assets/fonts/LeagueMono-VF.ttf");
```
```suggestion
@font-face {
  font-family: "FireSans";
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Regular.woff2");
```

Use a completely different font for fonts outside of 300 to 700 so people can see that a different font was actually used.

}

p {

>   font-family: "League Mono", "League Mono-complete", serif;

Reviewer: ```suggestion
  font-family: "League Mono", "FireSans", serif;
```

  font-size: 1.5rem;
}

p.one {
  font-weight: 200;
}

p.two {
  font-weight: 400;
}

p.three {
  font-weight: 600;
}

p.four {
  font-weight: 800;
}
```

#### Result

The result of this is:

> - Setting the `font-weight` property to `100` for "League Mono" gets a weight of 300.
> - Setting the `font-weight` property to `900` for "League Mono" gets a weight of 700.
> - Setting the `font-weight` property to `100` for "League Mono-complete" gets a weight of 100.
> - Setting the `font-weight` property to `900` for "League Mono-complete" gets a weight of 900.

Reviewer: ```suggestion
- Setting the `font-weight` property to `200` gets a `normal` version of FireSans.
- Setting the `font-weight` property to `400` gets "League Mono" at a weight of 400.
- Setting the `font-weight` property to `600` gets "League Mono" at a weight of 600.
- Setting the `font-weight` property to `800` gets a boldened version of "FireSans".
```

  font-size: 2rem;
}

p.one {
  font-weight: normal;
}

p.two {
  font-weight: bold;
}
```

#### Result

{{embedlivesample("Selecting normal and bold fonts", "", 300)}}

### Setting a range for a variable font

In this example we include a variable font, ["League Mono"](https://www.theleagueofmoveabletype.com/league-mono), using a single `@font-face` at-rule. We've used a `font-weight: 300 700` value to effectively limit the range of weights that are available. If a CSS rule uses our "League Mono" font, then if it specifies a weight outside this range the weight it gets is clamped to the range.

To show the effect of this we've included another font using "League Mono" that does not set the `font-weight` descriptor, and we've called this "League Mono-complete".

#### HTML

```html
<p class="one">League Mono-complete, font-weight 200</p>
<p class="two">League Mono, font-weight 400</p>
<p class="three">League Mono, font-weight 600</p>
<p class="four">League Mono-complete, font-weight 800</p>
```

#### CSS

```css
@font-face {
  font-family: "League Mono";
  src: url("https://mdn.github.io/shared-assets/fonts/LeagueMono-VF.ttf");
  font-weight: 300 700;
}

> @font-face {

Reviewer: ```suggestion
```

## Examples

### Selecting normal and bold fonts

In this example we include two fonts, one normal weight, one bold weight, from the ["Karantina"](https://fonts.google.com/specimen/Karantina) font family using two `@font-face` at rules. We set `font-weight` descriptors to match the weight of the fonts.

After this, CSS rules can select the normal or the bold font for the "Karantina" family just by setting the {{cssxref("font-weight")}} property. Note that the {{htmlelement("strong")}} HTML element also selects the bold font, because by default `<strong>` elements use a bold font.

#### HTML

```html
<p class="one">Karantina, normal weight paragraph</p>
<p class="two">Karantina, bold weight paragraph</p>
<strong>Karantina, default weight &lt;strong&gt; element</strong>
```

#### CSS

```css
@font-face {
  font-family: "Karantina";
  font-weight: normal;
  src: url("Karantina-Regular.woff2");
}

@font-face {
  font-family: "Karantina";
  font-weight: bold;
  src: url("Karantina-Bold.woff2");
}

body {
  font-family: "Karantina", serif;
  font-size: 3rem;
}

p.one {
  font-weight: normal;
}

p.two {

>   font-weight: bold;

Author: I don't understand this comment.

>  it's the default action expected of "font-weight: bold;"

I'm not sure what "the default action" means here. Like, if you don't set the font-weight descriptor, like this:

```css
@font-face {
  font-family: "Karantina";
  src: url("Karantina-Regular.woff2");
}

@font-face {
  font-family: "Karantina";
  src: url("Karantina-Bold.woff2");
}

body {
  font-family: "Karantina", serif;
  font-size: 3rem;
}

p.one {
  font-weight: normal;
}

p.two {
  font-weight: bold;
}
```

...then both paras and the `<strong>` will select "Karantina-Bold" because that's matched last (I think). The browser will also apply synthetic boldification to para 2 and the `<strong>`:

<img width="602" alt="Screen Shot 2024-04-03 at 8 03 40 PM" src="https://github.com/mdn/content/assets/432915/e3b45e63-71ba-48ce-8442-00726e243b2c">

It's only because we've set the `font-weight` descriptor that:
- para 1 selects "Karantina-Regular",
- para 2 and the `<strong>` select "Karantina-Bold" don't apply synthetic boldification
... and we get proper typography:

<img width="679" alt="Screen Shot 2024-04-03 at 8 05 16 PM" src="https://github.com/mdn/content/assets/432915/47233fd0-9e58-4842-ba71-dd87ebe5648e">

That's what this example is showing: by setting the font-weight descriptor, you enable a rule to select the appropriate font file by setting the font-weight property.

At least such is my understanding.

>   font-weight: bold;

Reviewer: Everything you wrote is correct. The issue is that the user will not be able to see a difference between a synthetic bold and the use of the actual bold font. 

I found this old example I wrote that barely works, but explains what I mean. At https://estelle.github.io/CSS/fonts/index.html#slide60 (slide 60 if the presentation doesn't actually move forward), if you change the font-weight in the `p.fwd` block to 700, you'll note the font family changes. The deck is old, and i had all the other fonts on my local machine when i did the presentation, so only the boldest font will work on your device. In that CSS, all the `@font` declarations are using the same `font-family` descriptor, but include different `font-weight` values and import complete different fonts. That was the user can see that a different font file is being used when the font-weight leads to different font-files used to render content. This is obviously not something one would do in production, but it enables the learner to really see what is going on.

In the current example on this page, the example will still be bold (or light) if the fonts don't even load. Yes, the second paragraph and the `<strong>` select `"Karantina-Bold"` and don't apply a synthetic a bold because we say so, but is it really just applying a synthetic bold? If we use drastically different fonts for normal v. bold in our `@font` `src` descriptor, the user will definitely see the effect. My suggestion was to invert bold and normal to show the difference, but drastically different fonts would be better (as it would be less confusing that "messing up" and explaining it)

>   font-weight: bold;

Author: I think this example should show the actual proper common usage of the font-weight descriptor to support multiple fonts in the same family, which is what it currently does. It implies "this is a good-practice way to use font-weight, if you're not using variable fonts". That's why I don't think we should change it to show something unrealistic, even if it is more illustrative.

I do agree that the counterfactual is not that clear, and that font synthesis makes the story more complicated. It might help to:

(a) find a font family that has a clearer differentiation between bold and normal weight
(b) maybe talk about the counterfactual (and even show it in the example) to explain what is happening (and that it is in fact broken)

That would make the example quite complex. I think one underlying problem here is that a lot of this explanation would go into a guide page, and because we don't have one, it is having to go here instead.

But I do think it is important for the first example here to show an actual good practice usage. Otherwise people have to infer it from unrealistic examples, and that's also putting a burden on them.

>   font-weight: bold;

Reviewer: All valid points. We really do need a font guide!

We can leave as is, but I did think of other options. Sharing in case you want to include one of them:

1) Duplicating the first example as a second example, and just inverting the bold / normal descriptors, with something like "This example "accidentally" inverts the `font-weight` descriptors, leading to a bold font being rendered when the `font-weight` property is `normal` and a normal font is rendered when the property is `bold`.

2) including 3 katerina fonts, serving 'light' up to 500, regular from 501 to 700, and bold 701 and above (though this would be for the numeric example after this one)

3)  in the numeric example, leaving gaps, so fallback font-family is used below 200 and above 800 or something similar.

}
```

#### Result

{{embedlivesample("Selecting normal and bold fonts", "", 300)}}

### Setting a range for a variable font

> In this example we include a variable font, ["Montserrat"](https://fonts.google.com/specimen/Montserrat), using a single `@font-face` at-rule. We've used a `font-weight: 300 700` value to effectively limit the range of weights that are available. If a CSS rule uses our "Montserrat" font, then if it specifies a weight outside this range the weight it gets is clamped to the range.

Reviewer: ```suggestion
In this example we include a variable font, ["Montserrat"](https://fonts.google.com/specimen/Montserrat), using a single `@font-face` at-rule. We've used a `font-weight: 300 700` value to effectively limit the range of weights for which this font will be used. If a CSS rule includes "Montserrat" as its first declared `font-family`, but the `font-weight` property value is outside of the weight range specified by the `font-weight` descriptor, the next font in the `font-family` list will be used, as the `Monserrat` font is clamped to the limited range specified by the descriptor.
```
this is what we want to say, though this might not be exactly how you want to say it.

> To show the effect of this we've included another font using "Montserrat" that does not set the `font-weight` descriptor, and we've called this "Montserrat-complete".

Reviewer: do you want to change the value to `auto`?

#### HTML

```html

> <p class="one">Montserrat, font-weight 100</p>
> <p class="two">Montserrat, font-weight 900</p>
> <p class="three">Montserrat-complete, font-weight 100</p>
> <p class="four">Montserrat-complete, font-weight 900</p>

Reviewer: ```suggestion
<p class="one">font-weight 200</p>
<p class="two">font-weight 400</p>
<p class="three">font-weight 600</p>
<p class="four">font-weight 800</p>
```

If we change to this, the two middle ones will be Monserrat and the first and list will be M-Complete. If we do this, however, we may want to use two very different fonts so the reader can quickly see that the middle two are different from the first and last. It is also how developers should be using `@font`; though most don't because font descriptors are rarely fully explained. In other words, I am super glad we are writing this!!!!!

> <p class="one">League Mono, font-weight 100</p>
> <p class="two">League Mono, font-weight 900</p>
> <p class="three">League Mono-complete, font-weight 100</p>
> <p class="four">League Mono-complete, font-weight 900</p>

Reviewer: ```suggestion
<p class="one">League Mono-complete, font-weight 200</p>
<p class="two">League Mono, font-weight 400</p>
<p class="three">League Mono, font-weight 600</p>
<p class="four">League Mono-complete, font-weight 800</p>
```

```

#### CSS

```css
@font-face {
  font-family: "Montserrat";
  src: url("Montserrat[wght].woff2");
  font-weight: 300 700;
}

@font-face {
  font-family: "Montserrat-complete";
  src: url("Montserrat[wght].woff2");
}

p {

>   font-size: 2rem;

Reviewer: ```suggestion
  font-size: 2rem;
  font-family: "Montserrat", "Montserrat-complete", serif;
```

}

> p.one {
>   font-family: "Montserrat", serif;
>   font-weight: 100;
> }
> 
> p.two {
>   font-family: "Montserrat", serif;
>   font-weight: 900;
> }
> 
> p.three {
>   font-family: "Montserrat-complete", serif;
>   font-weight: 100;
> }
> 
> p.four {
>   font-family: "Montserrat-complete", serif;
>   font-weight: 900;
> }

Reviewer: ```suggestion
p.one {
  font-weight: 100;
}

p.two {
  font-weight: 900;
}

p.three {
  font-weight: 100;
}

p.four {
  font-weight: 900;
}
```

> p {
>   font-size: 1.5rem;
> }
> 
> p.one {
>   font-family: "League Mono", serif;
>   font-weight: 100;
> }
> 
> p.two {
>   font-family: "League Mono", serif;
>   font-weight: 900;
> }
> 
> p.three {
>   font-family: "League Mono-complete", serif;
>   font-weight: 100;
> }
> 
> p.four {
>   font-family: "League Mono-complete", serif;
>   font-weight: 900;
> }

Reviewer: We want to show that you can use a series of fonts and the browser will select the most appropriate font for the job. We want users to see using a font family list, and not having to declare a font family every time a font-weight is changed. Hence, my recommendation is to change the  example so all paragraphs use the same font-family declaration, demonstrating that when the font-weight property value is withing the range of font-weight descriptor, that font will be used. In this case, 2 and 3 will get league mono and 1 and 4 will get league mono complete.

```suggestion
p {
  font-family: "League Mono", "League Mono-complete", serif;
  font-size: 1.5rem;
}

p.one {
  font-weight: 200;
}

p.two {
  font-weight: 400;
}

p.three {
  font-weight: 600;
}

p.four {
  font-weight: 800;
}
```

```

#### Result

The result of this is:

> - Setting the `font-weight` property to `100` for "Montserrat" gets a weight of 300.
> - Setting the `font-weight` property to `900` for "Montserrat" gets a weight of 700.
> - Setting the `font-weight` property to `100` for "Montserrat-complete" gets a weight of 100.
> - Setting the `font-weight` property to `900` for "Montserrat-complete" gets a weight of 900.

Reviewer: ```suggestion
- Setting the `font-weight` property to `200` for "Montserrat-complete" gets a weight of 200.
- Setting the `font-weight` property to `400` for "Montserrat" gets a weight of 400.
- Setting the `font-weight` property to `600` for "Montserrat" gets a weight of 600.
- Setting the `font-weight` property to `800` for "Montserrat-complete" gets a weight of 800.
```
